libs/langchain/langchain/chains/combine_documents/map_reduce.py

```diff
@@ -2,18 +2,266 @@
+
+            llm = ChatOpenAI(model="gpt-3.5-turbo")
+            extract_prompt = ChatPromptTemplate.from_template(
+                [
+                    ("system", "Given a user question, extract the most relevant parts of the following context:\n\n{context}"),
+                    ("human", "{question}"),
+                ]
+            )
+            map_documents_chain = create_map_documents_chain(llm, extract_prompt)
+
+            docs = [
+                Document(page_content="Jesse loves red but not yellow"),
+                Document(page_content = "Jamal loves green but not as much as he loves orange")
+            ]
+
+            map_documents_chain.invoke({"context": docs, "question": "Who loves green?"})
+    """  # noqa: E501
+    _document_prompt = document_prompt or PromptTemplate.from_template("{page_content}")
+
+    def _format_document(inputs: dict) -> str:
+        return format_document(inputs[DOCUMENTS_KEY], _document_prompt)
+
+    map_content_chain = (
+        RunnablePassthrough.assign(**{DOCUMENTS_KEY: _format_document}).with_config(
+            run_name="format_document"
+        )
+        | prompt
+        | llm
+        | StrOutputParser()
+    ).with_config(run_name="map_content")
+
+    map_doc_chain = (
+        RunnablePassthrough.assign(page_content=map_content_chain).with_config(
+            run_name="return_mapped_content"
+        )
+        | RunnableLambda(
+            lambda x: Document(
+                page_content=x["page_content"], metadata=x[DOCUMENTS_KEY].metadata
+            )
+        ).with_config(run_name="compile_document")
+    ).with_config(run_name="map_document")
+
+    def list_inputs(inputs: dict) -> list:
+        docs = inputs.pop(DOCUMENTS_KEY)
+        return [{DOCUMENTS_KEY: doc, **inputs} for doc in docs]
+
+    return (list_inputs | (map_doc_chain.map())).with_config(
+        run_name="map_documents_chain"
+    )
+
+
+def create_collapse_documents_chain(
+    llm: LanguageModelLike,
+    prompt: BasePromptTemplate,
+    *,
+    document_prompt: Optional[BasePromptTemplate] = None,
+    document_separator: str = DEFAULT_DOCUMENT_SEPARATOR,
+    token_max: int = 4000,
+    token_len_func: Optional[Callable[[str], int]] = None,
+) -> Runnable:
+    """Create
+
+    Args:
+
+    Returns:
+        An LCEL `Runnable` chain.
+
+    Example:
+        .. code-block:: python
+
+    """
+    _document_prompt = document_prompt or PromptTemplate.from_template("{page_content}")
+    _token_len_func = token_len_func or getattr(llm, "get_num_tokens", len)
+
+    def _format_inputs(inputs: dict) -> dict:
+        docs = inputs[DOCUMENTS_KEY]
+        inputs[DOCUMENTS_KEY] = document_separator.join(
+            format_document(doc, _document_prompt) for doc in docs
+        )
+        return {k: v for k, v in inputs.items() if k in prompt.input_variables}
+
+    def _format_metadata(inputs: dict) -> dict:
```

Comment:
generally not a fan of nested functions, i would vote pull out but this doesnt seem super general so maybe leave as private